aea/helpers/async_utils.py

```diff
@@ -477,3 +458,206 @@ def __init__(self, getters: List[Tuple[Callable[[Any], None], Callable]]):
+        """
+        if loop and threaded:
+            raise ValueError(
+                "You can not set a loop in threaded mode, cause own loop will be created"
+            )
+        self._loop = loop
+        self._threaded = threaded
+        self._task: Optional[asyncio.Task] = None
+        self._thread: Optional[Thread] = None
+        self._completed_event: Optional[asyncio.Event] = None
+        self._got_result = False
+        self._was_cancelled = False
+        self._is_running: bool = False
+
+    def start(self) -> bool:
+        """
+        Start runnable.
+
+        :return: bool started or not.
+        """
+        if self._task and not self._task.done():
+            logger.debug("already running")
+            return False
+        self._is_running = False
+        self._got_result = False
+        self._set_loop()
+        self._completed_event = asyncio.Event(loop=self._loop)
+        self._was_cancelled = False
+        self._set_task()
+
+        if self._threaded:
+            self._thread = Thread(
+                target=self._loop.run_until_complete, args=[self._task]
+            )
+            self._thread.start()
+
+        return True
+
+    def _set_loop(self) -> None:
+        """Select and set loop."""
+        if self._threaded:
+            self._loop = asyncio.new_event_loop()
+        else:
+            try:
+                self._loop = self._loop or asyncio.get_event_loop()
+            except RuntimeError:
+                self._loop = asyncio.new_event_loop()
+                asyncio.set_event_loop(self._loop)
+
+    def _set_task(self) -> None:
+        """Create task."""
+        self._task = self._loop.create_task(self._run_wrapper())
+
+    async def _run_wrapper(self) -> None:
+        """Wrap run() method."""
+        if not self._completed_event:
+            raise ValueError("Start was not called!")
+
+        try:
+            with suppress(asyncio.CancelledError):
+                return await self.run()
+        finally:
+            self._loop.call_soon_threadsafe(self._completed_event.set)
+
+    @property
+    def is_running(self) -> bool:
+        """Get running state."""
+        return self._is_running
+
+    @abstractmethod
+    async def run(self) -> Any:
+        """Implement run logic respectfull to CancelError on termination."""
+
+    def wait_completed(
+        self, sync: bool = False, timeout: float = None, force_result: bool = False
+    ) -> Awaitable:
+        """
+        Wait runnable execution completed.
+
+        :param sync: bool. blocking wait
+        :param timeout: float seconds
+        :param force_result: check result even it was waited.
+
+        :return: awaitable if sync is False, otherise None
+        """
+        if not self._task:
+            logger.warning("Runnable is not no started")
+            return ready_future
+
+        if self._got_result and not force_result:
+            logger.warning("Already got result, skip")
```

Comment:
Out of context this log statement will be hard to understand

Reply:
removed, used for debug